src/cow-react/modules/limitOrders/containers/RateInput/index.tsx

```diff
@@ -1,104 +1,75 @@
 import { RefreshCw } from 'react-feather'
 import { useAtomValue, useUpdateAtom } from 'jotai/utils'
-import { useWeb3React } from '@web3-react/core'
 
 import { HeadingText } from '@cow/modules/limitOrders/pure/RateInput/HeadingText'
 import { limitRateAtom, updateLimitRateAtom } from '@cow/modules/limitOrders/state/limitRateAtom'
-import { useCalculateRate } from '@cow/modules/limitOrders/hooks/useCalculateRate'
-import { useUpdateCurrencyAmount } from '@cow/modules/limitOrders/hooks/useUpdateCurrencyAmount'
 import { useLimitOrdersTradeState } from '@cow/modules/limitOrders/hooks/useLimitOrdersTradeState'
-import { useGetInitialPrice } from '@cow/modules/limitOrders/hooks/useGetInitialPrice'
-import usePrevious from 'hooks/usePrevious'
+import { toFraction } from '@cow/modules/limitOrders/utils/toFraction'
 
 export function RateInput() {
-  const { chainId } = useWeb3React()
-  const prevChainId = usePrevious(chainId)
-
-  // Price fetching
-  const { price: initialPrice } = useGetInitialPrice()
-
-  // Rate and currency amount hooks
-  const calculateRate = useCalculateRate()
-  const updateCurrencyAmount = useUpdateCurrencyAmount()
-
   // Rate state
-  const { isInversed, activeRate, isLoading, marketRate } = useAtomValue(limitRateAtom)
+  const { isInversed, activeRate, isLoading, executionRate, isLoadingExecutionRate, typedValue, isTypedValue } =
+    useAtomValue(limitRateAtom)
   const updateLimitRateState = useUpdateAtom(updateLimitRateAtom)
-  const prevIsInversed = usePrevious(isInversed)
 
   // Limit order state
-  const { inputCurrency, outputCurrency, inputCurrencyAmount } = useLimitOrdersTradeState()
+  const { inputCurrency, outputCurrency } = useLimitOrdersTradeState()
+  const areBothCurrencies = !!inputCurrency && !!outputCurrency
   const inputCurrencyId = inputCurrency?.symbol
   const outputCurrencyId = outputCurrency?.symbol
 
   const primaryCurrency = isInversed ? outputCurrencyId : inputCurrencyId
   const secondaryCurrency = isInversed ? inputCurrencyId : outputCurrencyId
 
+  // Handle rate display
+  const displayedRate = useMemo(() => {
+    if (isTypedValue) return typedValue || ''
+    else if (!activeRate || !areBothCurrencies || activeRate.equalTo(0)) return ''
+
+    const rate = isInversed ? activeRate.invert() : activeRate
+
+    return rate.toSignificant(6)
+  }, [activeRate, areBothCurrencies, isInversed, isTypedValue, typedValue])
+
   // Handle set market price
   const handleSetMarketPrice = useCallback(() => {
-    updateLimitRateState({ activeRate: marketRate })
-  }, [marketRate, updateLimitRateState])
+    updateLimitRateState({ activeRate: executionRate, isTypedValue: false })
+  }, [executionRate, updateLimitRateState])
 
   // Handle rate input
   const handleUserInput = useCallback(
     (typedValue: string) => {
-      updateLimitRateState({ activeRate: typedValue })
+      updateLimitRateState({ typedValue, activeRate: toFraction(typedValue, isInversed), isTypedValue: true })
```

Comment:
I love to see that now, when we parse the amount, we take into account if the user has the `isInverted` flag changed to invert what they type. 

If I understand it correctly, you are now storing the rate in the same units, and we don't invert it when we flip `isInversed`, but we just take it into account when we represent it.

Reply:
We do take it into account when its represented, but we also have to take it into account when user it typing value and `isInversed` is true.